ImputationPipeline/ScoringPart.wdl

```diff
@@ -229,6 +254,156 @@ task ScoreVcf {
+		Int block_buffer=10000000
+	}
+
+	Int disk_space =  3*ceil(size(vcf, "GB")) + 20
+
+	command <<<
+
+		tabix ~{vcf}
+
+		python3 << "EOF"
+		from cyvcf2 import VCF
+		import pandas as pd
+
+		vcf = VCF("~{vcf}", lazy=True)
+		samples = vcf.samples
+
+		def read_as_float(s):
+			try:
+				return float(s)
+			except ValueError:
+				pass
+
+		def add_allele_to_count(site, allele, dictionary):
+			if site in dictionary:
+				dictionary[site][allele]=[0]*len(samples)
+			else:
+				dictionary[site]={allele:[0]*len(samples)}
+
+		interactions_allele_counts = dict()
+		interactions_dict = dict()
+		positions = set()
+		if ~{if defined(sites) then "True" else "False"}:
+			with open("~{sites}") as f_sites:
+				sites = {s.strip() for s in f_sites}
+		else:
+			sites = {}
+		with open("~{interaction_weights}") as f:
+			for line in f:
+				if line.strip():
+					line_split = line.split()
+					if weight := read_as_float(line_split[8]):
+						if len(sites) == 0 or line_split[0] in sites and line_split[4] in sites:
+							add_allele_to_count(line_split[0], line_split[3], interactions_allele_counts)
+							add_allele_to_count(line_split[4], line_split[7], interactions_allele_counts)
+							interactions_dict[(line_split[0], line_split[3], line_split[4], line_split[7])]=weight
+							positions.add((line_split[1], int(line_split[2])))
+							positions.add((line_split[5], int(line_split[6])))
+
+		def add_self_exclusive_site(site, allele, dictionary):
+			if site in dictionary:
+				dictionary[site].add(allele)
+			else:
+				dictionary[site]={allele}
+
+		self_exclusive_sites = dict()
+		max_self_exclusive_sites = ~{if (defined(self_exclusive_sites)) then select_first([self_exclusive_sites]).maxAllowed else 0}
+		self_exclusive_sites_counts = [0]*len(samples)
+		if ~{if (defined(self_exclusive_sites)) then "True" else "False"}:
+			with open("~{select_first([self_exclusive_sites]).sites}") as f_self_exclusive_sites:
+				for line in f_self_exclusive_sites:
+					if line.strip():
+						line_split = line.split()
+						add_self_exclusive_site(line_split[0], line_split[3], self_exclusive_sites)
+						positions.add((line_split[1], int(line_split[2])))
+
+		#select blocks to read
+		positions = sorted(positions)
+		current_chrom=positions[0][0]
+		current_start=positions[0][1]
+		current_end = current_start+1
+		buffer=~{block_buffer}
+
+		blocks_to_read=[]
+		for site in positions:
+			if site[0] != current_chrom or site[1] - current_end > buffer:
+				blocks_to_read.append(current_chrom + ":" + str(current_start) + "-" + str(current_end))
+				current_chrom=site[0]
+				current_start=site[1]
+				current_end = current_start+1
+			else:
+				current_end = site[1] + 1
+
+		#last block
+		blocks_to_read.append(current_chrom + ":" + str(current_start) + "-" + str(current_end))
+
+		#count interaction alleles for each sample
+		for block in blocks_to_read:
+			for variant in vcf(block):
+				alleles = [a for a_l in [[variant.REF], variant.ALT] for a in a_l]
```

Comment:
I feel like there must be a more comprehensible way to write this section, but since it does what it's supposed to do let's leave it!